SlotMachine.hs

```diff
@@ -99,22 +99,19 @@ prettyGrid gen slots = putStr grid
     (newGrid, next_gen) = randomGrid gen slots
 
 putSpaces :: String -> String
-putSpaces [] = ""
-putSpaces (c:x) = [c] ++ "  " ++ putSpaces x
+putSpaces = concatMap (: "  ")
 
 randomGrid :: StdGen -> SlotMachine -> (String, StdGen)
-randomGrid gen slot = generateRandomGrid 15 "" gen normalizedSlots
+randomGrid gen slot = generate 15 gen
  where
+    
     normalizedSlots = normalizeSymbols (symbols slot)
-
--- Generates a character chain of random symbols
-generateRandomGrid :: Int -> String -> StdGen -> [(Char, Double)] -> (String, StdGen)
-generateRandomGrid len chain gen slots
-    | length chain == len = (chain, gen)
-    | otherwise = (chain ++ next_chain ++ rem_chain, next_gen)
-    where
-        (rem_chain, _) = generateRandomGrid (len-1) "" next_gen slots
-        (next_chain, next_gen) = randomSymbol gen slots
+    
+    generate 0 g = ([], g)
+    generate n g = (sym ++ chain, g') 
+        where
+            (sym, g') = randomSymbol g normalizedSlots
+            (chain, g'') = generate (n-1) g'
```

Comment:
significantly clearer code. also reduces the number of concatenations as well as removing the need to call `length`. Moving `generate` into the where clause allows us to use `normalizedSlots` without having to pass around a parameter to all of the recursive calls